File: ProjectEuler/051-100/S65.cs
using System.Linq;
using System.Numerics;

namespace ProjectEuler
{
    public class S65 : ISolution
    {
        public string GetAnswer()
        {
            BigInteger[] numerators = new BigInteger[101];
            BigInteger[] denominators = new BigInteger[101];

            numerators[1] = 2;
            numerators[2] = 3;
            denominators[1] = 1;
            denominators[2] = 1;

            for (int i = 3; i < 101; i++)
            {
                if (i % 3 == 0)
                {
                    numerators[i] = numerators[i - 1] * i * 2 / 3 + numerators[i - 2];
                    denominators[i] = denominators[i - 1] * i * 2 / 3 + denominators[i - 2];
                }
                else
                {
                    numerators[i] = numerators[i - 1] + numerators[i - 2];
                    denominators[i] = denominators[i - 1] + denominators[i - 2];
                }
            }

            return numerators[100].ToString().Select(c => c - '0').Sum().ToString();
        }
    }
}

File: src/051-100/S65.cc
#include "S65.h"
#include "../Utils.h"

using namespace std;

namespace ProjectEuler
{
std::string S65::GetAnswer()
{
	BigInt h2(0);
	BigInt h1(1);
	BigInt h(0);

	for (int i = 0; i < 100; i++)
	{
		long long a;
		if (i == 0)
		{
			a = 2;
		}
		else
		{
			if (i % 3 == 2)
			{
				a = 2 * (i / 3 + 1);
			}
			else
			{
				a = 1;
			}
		}

		BigInt bigA(a);
		h = bigA * h1 + h2;
		h2 = h1;
		h1 = h;
	}

	string s = h.toString();
	int sum = 0;
	for (char c : s)
	{
		sum += c - '0';
	}

	return std::to_string(sum);
}
} // namespace ProjectEuler
